Hercules Inpulse 200 (update March 2020)

DJPhatso opened the pull request:
Modified mapping files following the approval of the Inpulse 300

DJPhatso committed:
Add files via upload

Initial release of MIDI configuration files for Hercules DJControl Inpulse 200

DJPhatso committed:
Add files via upload

DJPhatso committed:
Hercules-Inpulse-200

DJPhatso committed:
Delete Hercules-DJControl-Inpulse-200-script.js

Deleted old file

DJPhatso committed:
Delete Hercules_DJControl_Inpulse_200.midi.xml

Deleted old file

DJPhatso committed:
Update Hercules-DJControl-Inpulse-200-script.js

Minor corrections that eslint didn't catch.

Holzhaus commented:
Hey @DJPhatso, thanks for contributing yet another controller mapping. Can you please rebase on 2.2?

Holzhaus commented:
https://mixxx.org/wiki/doku.php/using_git#targeting_another_base_branch

Holzhaus commented:
Our pre-commit-pr job on Travis CI found some style issues:
https://travis-ci.org/github/mixxxdj/mixxx/jobs/660633844

Can copy and paste the diff from the build log into a new file names `fix-issues.patch`, the run `git apply fix-issues.patch`? That should resolve the issues.

Here's the diff:giu
```diff
diff --git a/res/controllers/Hercules-DJControl-Inpulse-200-script.js b/res/controllers/Hercules-DJControl-Inpulse-200-script.js
index fa12bac36..2b8a349d7 100644
--- a/res/controllers/Hercules-DJControl-Inpulse-200-script.js
+++ b/res/controllers/Hercules-DJControl-Inpulse-200-script.js
@@ -50,8 +50,8 @@ DJCi200.init = function() {
     DJCi200.scratchButtonState = true;
     // Scratch Action
     DJCi200.scratchAction = {
-    1: DJCi200.kScratchActionNone,
-    2: DJCi200.kScratchActionNone
+        1: DJCi200.kScratchActionNone,
+        2: DJCi200.kScratchActionNone
     };
 
     //Turn On Vinyl buttons LED(one for each deck).
@@ -61,11 +61,11 @@ DJCi200.init = function() {
     //Turn On Browser button LED
     midi.sendShortMsg(0x90, 0x04, 0x05);
 
-   //Softtakeover for Pitch fader
-   engine.softTakeover("[Channel1]", "rate", true);
-   engine.softTakeover("[Channel2]", "rate", true);
-   engine.softTakeoverIgnoreNextValue("[Channel1]", "rate");
-   engine.softTakeoverIgnoreNextValue("[Channel2]", "rate");
+    //Softtakeover for Pitch fader
+    engine.softTakeover("[Channel1]", "rate", true);
+    engine.softTakeover("[Channel2]", "rate", true);
+    engine.softTakeoverIgnoreNextValue("[Channel1]", "rate");
+    engine.softTakeoverIgnoreNextValue("[Channel2]", "rate");
 
     //Set effects Levels - Dry/Wet
     engine.setParameter("[EffectRack1_EffectUnit1_Effect1]", "meta", 0.6);
@@ -77,7 +77,7 @@ DJCi200.init = function() {
     engine.setParameter("[EffectRack1_EffectUnit1]", "mix", 1);
     engine.setParameter("[EffectRack1_EffectUnit2]", "mix", 1);
 
-	// Ask the controller to send all current knob/slider values over MIDI, which will update
+    // Ask the controller to send all current knob/slider values over MIDI, which will update
     // the corresponding GUI controls in MIXXX.
     midi.sendShortMsg(0xB0, 0x7F, 0x7F);
 };
diff --git a/res/controllers/Hercules_DJControl_Inpulse_200.midi.xml b/res/controllers/Hercules_DJControl_Inpulse_200.midi.xml
index 167611dd6..2c9b7634f 100644
--- a/res/controllers/Hercules_DJControl_Inpulse_200.midi.xml
+++ b/res/controllers/Hercules_DJControl_Inpulse_200.midi.xml
@@ -790,7 +790,7 @@
                     <normal />
                 </options>
             </control>
-            
+
             <!--Continuous Controllers (CC) - Faders/knobs -->
             <!-- CC's MIDI Channel 1 (0xB0 Standard mode)-->
             <!--Crossfader-->
@@ -1645,4 +1645,4 @@
             </output>
         </outputs>
     </controller>
-</MixxxMIDIPreset>
\ No newline at end of file
+</MixxxMIDIPreset>
```

DJPhatso commented:
> Can copy and paste the diff from the build log into a new file names `fix-issues.patch`, the run `git apply fix-issues.patch`? That should resolve the issues.
> 

Let's see if I understand correctly:

- I should create a new text file called `fix-issues.patch`, paste the build log into it and save it (in the root of my Mixxx folder i would guess?);

- Once done, run the command `git apply fix-issues.patch`.


So I tried that, and I got 'error: corrupt patch at line 61'

Holzhaus commented:
> > Can copy and paste the diff from the build log into a new file names `fix-issues.patch`, the run `git apply fix-issues.patch`? That should resolve the issues.
> 
> Let's see if I understand correctly:
> 
> * I should create a new text file called `fix-issues.patch`, paste the build log into it and save it (in the root of my Mixxx folder i would guess?);

Not the complete build log, just the diff that I copy-and-pasted from the build log to my comment above.

> * Once done, run the command `git apply fix-issues.patch`.
> 
> So I tried that, and I got 'error: corrupt patch at line 61'

You need to copy and paste the complete diff and not add, remove or modify any lines. It works fine for me.

After you've done that, you can delete the `fix-issues.patch` file and commit the changes.

DJPhatso commented:
Exactly what I had done. Created a new one just in case, tried again, same error
[fix-issues.zip](https://github.com/mixxxdj/mixxx/files/4324382/fix-issues.zip)

Holzhaus commented:
Maybe your editor changed the formatting automatically or something like this. In that case download (not copy&paste) this one: https://gist.github.com/Holzhaus/d1c1c10f1b48c1b22961e420c568a67c

DJPhatso commented:
No better

error: patch failed: res/controllers/Hercules-DJControl-Inpulse-200-script.js:61
error: res/controllers/Hercules-DJControl-Inpulse-200-script.js: patch does not apply

Holzhaus commented:
This patch definitely works fine. Did you change your branch locally (i.e. made changed to the controller script files)? If you didn't commit them yet, you can check with `git status`.

DJPhatso commented:
Haven't made any changes since it was committed and status tells me that it is indeed up to date.

Holzhaus committed:
Hercules DJControl Inpulse 200: Fix some code stye issues

Holzhaus commented:
That is super strange. It works fine for me. Anyway, checkout out DJPhatso/mixxx#6.

DJPhatso committed:
Merge pull request #6 from Holzhaus/Hercules-Inpulse-200

Hercules DJControl Inpulse 200: Fix some code stye issues

Holzhaus commented:
Ok, the code style issues seem to be fixed. Ignore the CI build fails - our CI for 2.2 does not work properly. I'll review soon.

Holzhaus reviewed (approved):
@DJPhatso, thanks for contributing this mapping. It's really similar to the Inpulse 300 mapping that was merged a few days ago, so this should be fine. @Be-ing @Swiftb0y Do any of you want to take a look or shall I merge?

Swiftb0y commented:
I'll take a look.

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-200-script.js:
```diff
@@ -0,0 +1,182 @@
+    engine.setParameter("[EffectRack1_EffectUnit2_Effect2]", "meta", 0.6);
+    engine.setParameter("[EffectRack1_EffectUnit2_Effect3]", "meta", 0.6);
+    engine.setParameter("[EffectRack1_EffectUnit1]", "mix", 1);
+    engine.setParameter("[EffectRack1_EffectUnit2]", "mix", 1);
```
This has nothing to do with the controller itself. afaik this should not be included in the official mapping. @Holzhaus opinions?

Holzhaus replied:
Yes, true. I vaguely remember pointing out similar code in one of the past controller mapping PRs, but I can't find it anymore. I don't know how we proceeded there.

Swiftb0y replied:
I have the same issue in #2546 multiple times. I think the primary issue is that people abuse the controller script for some quick automation in JS. IMO the  best solution to this issue would be, that we create a second JS environment that is controller independent and let people do some basic scripting by manipulating COs that way. There are some usecases that are very userspecific and can be implemented solely via JS and the current scripting capabilities via CO's. Due to them being so user specific, it wouldn't make sense to implement them in C++ and merge them into Mixxx but many users are also not tech savvy enough to maintain their own fork. It would make sense for them to have a dynamic scripting environment.
Should I create a Zulip topic to discuss this further?

Holzhaus replied:
> Should I create a Zulip topic to discuss this further?

Yup.

DJPhatso replied:
For info, I've used this method of pre-setting the effects level in the mapping #2370 and #2061 as well to help with the lack of FX controls (pot or encoders) on these controllers. The idea is that I can then map a button or a pad to simply enable/disable the fx.

I had tried having the button command setting both the fx On/Off and level at the same times, but that often didn't work properly. That was early in the new FX Frameworks so maybe things have changed but I'm sure my limited coding knowledge might also have come into play.

Anyway, if there is a better method of doing things, I'm all for it, and removing it is not that much of an issue since this controller includes a Filter knob (where the two others mentioned didn't.)

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-200-script.js:
```diff
@@ -0,0 +1,182 @@
+};
+
+DJCi200._convertWheelRotation = function(value) {
+    // When you rotate the jogwheel, the controller always sends either 0x1
+    // (clockwise) or 0x7F (counter clockwise). 0x1 should map to 1, 0x7F
+    // should map to -1 (IOW it's 7-bit signed).
```
Can you confirm this is the case at higher speeds as well? Controllers usually bundle messages together to not oversaturate the bus. The resulting packets then carry a higher value with them.

DJPhatso replied:
yes, this is the case for our (Hercules) controllers. Included a capture of scratching with the Inpulse200.
[Inpulse200_Jog_capture.txt](https://github.com/mixxxdj/mixxx/files/4771305/Inpulse200_Jog_capture.txt)

Holzhaus commented on res/controllers/Hercules_DJControl_Inpulse_200.midi.xml:
```diff
@@ -0,0 +1,1648 @@
+            <!--Sync-->
+            <control>
+                <group>[Channel1]</group>
+                <key>sync_master</key>
```
@DJPhatso Are you actually using this? If yes, can you explain what you use it for? We're reworking master sync behavior in PR #2376 and think about dropping support for setting `sync_master` manually. (ping @ywwg)

DJPhatso replied:
You mean personally ? No I don't, but I had seen it mapped on other controllers, so I figured it might be useful to someone. But given it is a "secondary" SHIFT function, you have to know that it is there.

Holzhaus replied:
Yeah, maybe it makes sense to map it to toggle quantize or trigger beatsync instead then.

DJPhatso replied:
Possible. Or remove it entirely. Then it would need to be changed in #2465 as well?

Holzhaus replied:
Yes, but we don't have a definitive decision regarding #2376 yet.

DJPhatso replied:
Any news on that ?

Holzhaus replied:
Sorry, it's still being discussed here: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/2.2E3.20planning.3A.20https.3A.2F.2Fgithub.2Ecom.2Fmixxxdj.2Fmixxx.2Fprojects.2F2/near/192204135

Holzhaus replied:
Looks like we keep the ability to set sync master explicitly, so it's up to you if you want to keep this to use the button for something else (like toggling quantize) .

DJPhatso replied:
Noted. I'll leave them as is now. 

Meanwhile, I made some changes to the the .xml to correct an issue I found and addressed an upcoming update to the device that will change how the pads behave in normal mode. This way, the mapping will work whether or not the users update their devices. I will add the relevant information in the description page once the update is released.

DJPhatso committed:
Changes related to FX pad 

- Corrected a bad assignment  to Effect Rack On/off
- Mapped FX On/Off to both standard and SHIFT mode for the FX pads (related to an upcoming update to the device that will change how the pads work ).

Holzhaus reviewed (approved):
LGTM, thank you! Sorry for the delay, I'll go ahead and merge now.

DJPhatso commented:
Thanks Jan.

Holzhaus commented:
@DJPhatso I just noticed that this controller is lacking documentation in the manual (probably because this was originally merged for 2.2.5, before we added controller documentation in the manual). You could make a PR to the mixxxdj/manual to add it? Something similar to the documentation for the [Pioneer DDJ-200](https://manual.mixxx.org/2.3/en/hardware/controllers/pioneer_ddj_200.html) or the [Roland DJ-505](https://manual.mixxx.org/2.3/en/hardware/controllers/roland_dj_505.html) would be great.

DJPhatso commented:
Hi Jan,

I had created a page for the Inpulse 500 (like I did for the others Hercules devises I worked on) on the old documentation system:

https://github.com/<https://github.com/mixxxdj/mixxx/wiki/Hercules-Djcontrol-Inpulse-500>mixxxdj<https://github.com/mixxxdj/mixxx/wiki/Hercules-Djcontrol-Inpulse-500>/mixxx/wiki/Hercules-<https://github.com/mixxxdj/mixxx/wiki/Hercules-Djcontrol-Inpulse-500>Djcontrol<https://github.com/mixxxdj/mixxx/wiki/Hercules-Djcontrol-Inpulse-500>-<https://github.com/mixxxdj/mixxx/wiki/Hercules-Djcontrol-Inpulse-500>Inpulse<https://github.com/mixxxdj/mixxx/wiki/Hercules-Djcontrol-Inpulse-500>-500<https://github.com/mixxxdj/mixxx/wiki/Hercules-Djcontrol-Inpulse-500>

Can it be imported to the new manual as the other seemingly did?

Cheers

Ben

Ben (sent from mobile)


________________________________
From: Jan Holthuis ***@***.***>
Sent: Thursday, May 13, 2021 12:32:53 PM
To: mixxxdj/mixxx ***@***.***>
Cc: Benoit Duval ***@***.***>; Mention ***@***.***>
Subject: Re: [mixxxdj/mixxx] Hercules Inpulse 200 (update March 2020) (#2542)


@DJPhatso<https://github.com/DJPhatso> I just noticed that this controller is lacking documentation in the manual (probably because this was originally merged for 2.2.5, before we added controller documentation in the manual). You could make a PR to the mixxxdj/manual to add it? Something similar to the documentation for the Pioneer DDJ-200<https://manual.mixxx.org/2.3/en/hardware/controllers/pioneer_ddj_200.html> or the Roland DJ-505<https://manual.mixxx.org/2.3/en/hardware/controllers/roland_dj_505.html> would be great.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/mixxxdj/mixxx/pull/2542#issuecomment-840678687>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALQYI6O7Q6EHZTUTR7M55VLTNP5LLANCNFSM4LE67IFA>.
